### PR TITLE
a new method tryGetModel() that call cb on error or undefined

### PR DESCRIPTION
**Pull request by haoadoreorange**
Signed-off-by: Quoc-Hao Tran <quoc-hao.tran@polymtl.ca>

### CODE REVIEW

**Comment by haoadoresorange** [comment]
I just find myself rewriting a lot of Error handler for this use case when getting the response model, so this is a suggestion, I'm not sure if it's the appropriate solution.

File: tsp-typescript-client/src/protocol/tsp-client-response.ts [hunk 1]
```diff
@@ -32,6 +32,15 @@ export class TspClientResponse<T> {
     public getModel(): T | undefined {
         return this.responseModel;
     }
+    
+    /**
+     * Get the model from the server, or throw custom Error
+     */
+    public tryGetModel(CustomError: new (msg: string, code?: number) => Error): never | T {
```
Comment (paul-marechal) [inline tsp-typescript-client/src/protocol/tsp-client-response.ts:39]: `never` actually is the empty type. Here the return type is strictly equivalent to just `T`.

**Review by paul-marechal — COMMENTED** [review]
I understand the feeling, but I'm not sure if this is a very common usage scenario for this API... But maybe I'm wrong, I'd like to hear more feedback.

In the meantime, nothing prevents you from implementing those utility functions in your own project for reuse.

File: tsp-typescript-client/src/protocol/tsp-client-response.ts (hunk 1 above)
Comment (haoadoresorange) [reply tsp-typescript-client/src/protocol/tsp-client-response.ts:39]: Yes that's true ! I Sorry I was just quickly adding it as an example, not actually using it. Thanks for spotting it out.

**Comment by haoadoresorange** [comment]
Hi, sorry for the late reply !

> Is this change for being able to re-use a common error handler implementation? Please let me know what the advantages are

Yes, or just pass a custom error. It was just a short way for me to write the code, so instead of 
```typescript
const response = tsp.fetchExperiments();
try{
   if (! response.isOk()) throw new CustomError(response.statusMessage, response.statusCode)
   if (! response.getModel()) throw new CustomError(response.statusMessage)
} catch (e) {
   handleError(e)
}
```
Now I write
```typescript
   response.tryGetModel(handleError)
```
Idk about how the `isOk` and `getModel` are usually used, but I find myself just couple them together most of the time. Are you familiar with the matter ?

**Comment by haoadoresorange** [comment]
should I squash all the commits ?

**Review by bhufmann — APPROVED** [review]
I'm ok with this new API. Thanks!

**Comment by marco-miller** [comment]
@bhufmann -Bernd, how should we handle this failing legal check you think?

**Comment by bhufmann** [comment]
> @bhufmann -Bernd, how should we handle this failing legal check you think?

So that @haoadoreorange can contribute this project, he has to sign the ECA as described in the [contribution guide: ](https://github.com/eclipse-cdt-cloud/tsp-typescript-client/blob/master/CONTRIBUTING.md) and then update the patch (e.g. rebase).

**Review by marco-miller — CHANGES_REQUESTED** [review]
-As per the previous comments thread (feedback to/for the author; thanks again).

**Comment by bhufmann** [comment]
Closing this PR since it hasn't be active for a while. Please reopen if required.

**Pull request closed without merging**